pinot-transport/src/main/java/com/linkedin/pinot/requestHandler/BrokerRequestHandler.java

```diff
@@ -634,24 +555,51 @@ private BrokerResponse processFederatedBrokerRequest(@Nonnull BrokerRequest orig
   }
 
   /**
+   * Gather responses from servers, append processing exceptions to the processing exception list passed in.
+   *
+   * @param compositeFuture composite future returned from scatter phase.
+   * @param scatterGatherStats scatter-gather statistics.
+   * @param isOfflineTable whether the scatter-gather target is an OFFLINE table.
+   * @param tableName table name.
+   * @param processingExceptions list of processing exceptions.
+   * @return server response map.
+   */
+  @Nullable
+  private Map<ServerInstance, ByteBuf> gatherServerResponses(
+      @Nonnull CompositeFuture<ServerInstance, ByteBuf> compositeFuture, @Nonnull ScatterGatherStats scatterGatherStats,
+      boolean isOfflineTable, @Nonnull String tableName, @Nonnull List<ProcessingException> processingExceptions) {
+    try {
+      Map<ServerInstance, ByteBuf> serverResponseMap = compositeFuture.get();
+      Map<String, Long> responseTimes = compositeFuture.getResponseTimes();
+      scatterGatherStats.setResponseTimeMillis(responseTimes, isOfflineTable);
+      return serverResponseMap;
+    } catch (Exception e) {
+      LOGGER.error("Caught exception while fetching responses for table: {}", tableName, e);
+      _brokerMetrics.addMeteredTableValue(tableName, BrokerMeter.RESPONSE_FETCH_EXCEPTIONS, 1);
+      processingExceptions.add(QueryException.getException(QueryException.BROKER_GATHER_ERROR, e));
+      return null;
```

Comment:
maybe better to return an empty map and check for empty map outside, but your call.